libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,634 @@
+      s_store.soft_error_flag = true;
+    }  // Soft errors just set flag and continue
+  }
+}
+
+static void prv_init_fsm(void *context) {
+  fsm_init(&s_dgram_fsm, "can_dgram_fsm", &state_idle, context);
+  fsm_state_init(state_idle, prv_process_idle);
+  fsm_state_init(state_tx_protocol_version, prv_process_protocol_version_tx);
+  fsm_state_init(state_tx_crc, prv_process_crc_tx);
+  fsm_state_init(state_tx_dgram_type, prv_process_dgram_type_tx);
+  fsm_state_init(state_tx_dst_len, prv_process_dst_len_tx);
+  fsm_state_init(state_tx_dst, prv_process_dst_tx);
+  fsm_state_init(state_tx_data_len, prv_process_data_len_tx);
+  fsm_state_init(state_tx_data, prv_process_data_tx);
+
+  fsm_state_init(state_rx_protocol_version, prv_process_protocol_version_rx);
+  fsm_state_init(state_rx_crc, prv_process_crc_rx);
+  fsm_state_init(state_rx_dgram_type, prv_process_dgram_type_rx);
+  fsm_state_init(state_rx_dst_len, prv_process_dst_len_rx);
+  fsm_state_init(state_rx_dst, prv_process_dst_rx);
+  fsm_state_init(state_rx_data_len, prv_process_data_len_rx);
+  fsm_state_init(state_rx_data, prv_process_data_rx);
+
+  fsm_state_init(state_done, prv_process_done);
+}
+
+static void prv_rx_watchdog_expiry_cb(void *context) {
+  LOG_WARN("WATCHDOG EXPIRED. EXITING WITH ERROR...\n");
+  event_raise(s_store.error_event, HARD_ERROR);
+}
+
+StatusCode can_datagram_init(CanDatagramSettings *settings) {
+  // Populate storage
+  s_store.tx_event = settings->tx_event;
+  s_store.rx_event = settings->rx_event;
+  s_store.repeat_event = settings->repeat_event;
+  s_store.error_event = settings->error_event;
+
+  fifo_init(&s_fifo, s_buffer);
+  prv_init_fsm((void *)&s_store);
+  s_store.status = DATAGRAM_STATUS_IDLE;
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_datagram_rx(uint8_t *data, size_t len, bool start_message) {
+  if (len > CAN_BUFFER_SIZE) {
+    return STATUS_CODE_OUT_OF_RANGE;
+  }
+
+  // Only run if start_listener called
+  if (!s_store.rx_listener_enabled) {
+    return STATUS_CODE_UNINITIALIZED;
+  }
+  if (start_message) {
+    // Only process start messages if rx is not currently active
+    // or if soft error has occured
+    if (s_store.status != DATAGRAM_STATUS_ACTIVE || s_store.soft_error_flag) {
+      // Start frame rcv'd -> status now active
+      // Commence rx side of state machine
+      s_store.status = DATAGRAM_STATUS_ACTIVE;
+      s_store.soft_error_flag = false;
+      event_raise_no_data(s_store.rx_event);
+      watchdog_start(&s_watchdog, RX_WATCHDOG_TIMEOUT_MS, prv_rx_watchdog_expiry_cb, NULL);
+      return STATUS_CODE_OK;
+    } else {
+      return STATUS_CODE_UNREACHABLE;
+    }
+  } else if (s_store.status != DATAGRAM_STATUS_ACTIVE) {
+    // If start message not received, do not rx data
+    return STATUS_CODE_UNINITIALIZED;
+  }
+
+  // Rx normal data
+  watchdog_kick(&s_watchdog);
+  prv_write_rx_buffer(data, len);
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_datagram_start_listener(CanDatagramRxConfig *config) {
```

Comment:
What's the strategy for notifying client code when we've successfully received a datagram? Currently we seem to just notify by setting the status - maybe we should take an optional "rx success" callback? (And on that note - perhaps an optional "tx success" callback in `can_datagram_start_tx` and an optional "error" callback in `can_datagram_init` (or maybe both tx and start listener functions) for hard errors?)